Change LTI update to allow selecting multiple sets.

somiaj opened the pull request:
Update the LTI Mass Update page to allow updating either all users or a single user for multiple sets with a single update.

Alex-Jordan commented on lib/Mojolicious/WeBWorK/Tasks/LTIMassUpdate.pm:
```diff
@@ -39,10 +39,16 @@ sub run ($job, $userID = '', $setID = '') {
-		$updateUsers{$userID} = [$setID];
-	} elsif ($setID && $ce->{LTIGradeMode} eq 'homework') {
-		%updateUsers = map { $_ => [$setID] } $db->listSetUsers($setID);
+	if ($setIDs && ref($setIDs) eq 'ARRAY' && $ce->{LTIGradeMode} eq 'homework') {
```
All blocks except the last `else` fall under `$setIDs && $ce->{LTIGradeMode} eq 'homework'`, so this might be more human readable if they were wrapped in one conditional for that.

```
	if ($setIDs && $ce->{LTIGradeMode} eq 'homework') {
		if (ref($setIDs) eq 'ARRAY' && $userID) {
			$updateUsers{$userID} = $setIDs;
		} elsif (ref($setIDs) eq 'ARRAY') {
			%updateUsers = map { $_ => $setIDs } $db->listUsers;
		} elsif ($userID) {
			$updateUsers{$userID} = [$setIDs];
		} else {
			%updateUsers = map { $_ => [$setIDs] } $db->listSetUsers($setIDs);
		}
	} else {
	...
```

Alex-Jordan replied:
It might simplify the logic to precede all of this with:

```
$setIDs = [$setIDs] unless (ref $setIDs);
```

and just always act as if it is an array, even with just one entry.

Alex-Jordan reviewed (approved):
I have not actually tested, but once this is merged and I pull into production, I'll test there.

somiaj commented on lib/WeBWorK/ContentGenerator/Instructor/LTIUpdate.pm:
```diff
@@ -23,36 +23,32 @@ sub initialize ($c) {
-	my $setID       = $c->param('updateSetID');
-	my $userID      = $c->param('updateUserID');
-	my $prettySetID = format_set_name_display($setID // '');
+	my @setIDs  = ($c->param('updateSetID'));
+	my $userID  = $c->param('updateUserID');
+	my $allSets = $c->param('selectAllSets');
```
Looks like my logic was flawed, I was trying to use this to keep the old message of 'all sets queued'. Is there a way I can tell if all items were selected or not, for some reason I only though I got the list of what was selected in the multiple select, and was looking for a way to tell if everything was selected or not. I guess I could compare it to the number of possible sets.

somiaj commented:
Update via @drgrice1 review. I do forget how to do plural in maketext, for instance the messages may say `1 sets` vs `1 set`, since in all cases I chose to just list the number of sets. I could list all the set IDs, I just thought that might get a bit long, but if that is preferred I could change `3 sets`, to 'sets: id1, id2, id3', and list all the actual set IDs, though this may get long when all sets are selected (which was I decided to just list the number of sets).

drgrice1 commented:
By the way, aren't you supposed to still be out climbing mountains, and not working on WeBWorK development?

somiaj commented:
@drgrice1 Is there a way to get the list of set IDs assigned to all users put into a hash `userID => $setIDs`, with a single database call. This way I don't have to call the database for each user in a loop (what is currently being done in the LTIUpdate templates file) and also have access to the number of sets assigned to each user, so I can determine if I should say 'all sets' or '5 sets'. Basically I want to the put the following in a single call (or maybe two, one to get user list, and one to get set list for each user).

```
	my @allUsers    = $db->listUsers;
	my %allUserSets = map { $_ => [ sort $db->listUserSets($_) ] } @allUsers;
```

somiaj commented:
Mountains have been climbed, lots of southern utah has been seen, now at the air b and b with some time to work.

drgrice1 commented:
Wait.  That isn't right.  Just a second.

somiaj commented:
Oh wait, that is a single database call, missed that. Thanks.

somiaj commented:
@dlglin I think you need to restart the `webwork2-job-queue.service` after switching to this branch, as I am not able to reproduce what you are seeing. This is what I see when I select two sets:

```
Preparing to submit grade to LMS for user XXXXXXXX and set SETID1
LMS user id is not available for this user.
Preparing to submit grade to LMS for user XXXXXXXX and set SETID2
LMS user id is not available for this user.
```

I don't have fully functional LTI setup on my dev setup, so that second message is just failing due to that, but I am seeing both setIDs begin processed for all users in my logs in the "Job Manager".

dlglin commented:
Yup.  I forgot to restart the job queue.  It's working as expected now.

drgrice1 commented on lib/WeBWorK/DB.pm:
```diff
@@ -1388,7 +1388,7 @@ sub getUserSet {
 }
 
 sub getUserSets {
-	my ($self, @userSetIDs) = shift->checkArgsRefList(\@_, qw/user_id set_id/);
```
This is what is causing the issue that @dlglin noticed.  This shouldn't be changed.

somiaj commented:
@dlglin Could you retest things, I updated the logic. I removed testing if `$setIDs` is an array reference, since this function is only run in one place and it will always be an array reference, the logic now just assumes `$setIDs` is an array reference and no longer tests for that case.

I also updated the messaging a bit. If a single set is selected, I use the old message of 'LTI update of set SETID for user USERID queued' in that case, and only list the count of sets if more than one is selected.

drgrice1 commented on htdocs/js/LTIUpdate/ltiupdate.js:
```diff
@@ -1,33 +1,48 @@
+
+	// Deal with the select all sets checkbox to either select all sets or uncheck if a set is unselected.
+	userMenu?.addEventListener('change', updateSetMenu);
+	setMenu?.addEventListener('change', (e) => {
+		if (!e.target.selected) setSelectAll.checked = false;
 	});
```
This isn't working.  `e.target.selected` is always undefined.  So this might as well just always set `setSelectAll.checked = false`.  The problem is that `e.target` is always the select element, and not on of the options.  A select element does not have the `checked` property.  I suggest changing this to
```suggestion
	setMenu?.addEventListener('change', () => {
		let allSelected = true;
		for (const option of setMenu.options) {
			if (!option.selected) allSelected = false;
		}
		setSelectAll.checked = allSelected;
	});
```
With that the select all check will accurately reflect when all sets are selected or not.

drgrice1 reviewed (approved):
Other than the select all issue, I think this looks good.

drgrice1 commented:
I will merge this now.
